source/asynch_socket.c

```diff
@@ -533,12 +533,29 @@ static socket_error_t lwipv4_socket_bind(struct socket *sock, const struct socke
 }
 
 static void rx_core(struct socket * s, struct pbuf *p) {
-
+    /** NOTE: the receive callback for both UDP and TCP are responsible for deallocating pbufs, so there is no call to
+     *  pbuf_ref() required here.
+     */
     __disable_irq();
     if (s->rxBufChain == NULL) {
         s->rxBufChain = p;
     } else {
-        pbuf_cat((struct pbuf *) s->rxBufChain, p);
+        struct pbuf * q = (struct pbuf *)s->rxBufChain;
+        switch(s->family) {
+            case SOCKET_DGRAM:
+                // find the last element of the buffer chain.
+                while (q->next) {q = q->next;}
+                /**
+                 * Attach p to it wihout changing the tot_len of the buffer chain
+                 * NOTE: This is not how pbufs are intended to work, but it is necessary to deal with a) fragmentation
+                 * and b) packet queueing
+                 */
+                q->next = p;
```

Comment:
Some useful background information: `tot_len` is intended to contain the length of the current `pbuf` and every subsequent `pbuf`.
 
`pbuf_cat(p,q)` does the following:
* Increase `p->tot_len` and the `tot_len` of every linked `pbuf` by `q->tot_len`
* Find the last `pbuf` linked by `p`
* Link `q` to the last `pbuf` linked by `p`

`pbuf_chain` does `pbuf_cat` and `pbuf_ref`

Since `pbuf_chain` and `pbuf_cat` both modify `tot_len`, I don't believe that either of these is correct.